doc/pam_oidc.8
.TH PAM_OIDC 8 "28 August 2025" "pam_oidc" "OIDC PAM Module Manual"


.SH NAME
pam_oidc \- Module for OpenID Connect (OIDC) authentication


.SH SYNOPSIS
.B pam_oidc\&.so id_token
[\fB--openid-config-url=\fIconfig_url\fR]
.br
.B pam_oidc\&.so poll
.BI --client-id= client_id
.BI --client-secret= client_secret
[\fB--openid-config-url=\fIconfig_url\fR]
.br
.B pam_oidc\&.so local_auth
[\fB--verify-endpoint=\fIverify_endpoint\fR]


.SH DESCRIPTION
.B pam_oidc
is a PAM module that incorporates OIDC to your Linux login\&.
It features a few modes of operation\&.

In
.B id_token
mode, the user is prompted for OIDC ID token in JWT format\&.
Using JWK (JSON Web Key) Set
(that contains the public part of key used to sign the token)
obtained from OpenID Provider's dedicated endpoint,
the token's cryptographic signature is verified\&.
If successful, user is granted access if proper account is requested\&.
For now, only simple mapping is available:
the requested account's username must be equal to "user" string
+ the value of user's "sub" claim (e\&.g\&. "user101977677136526966383")\&.

In
.B poll
mode, the user is shown URL and code\&.
The user should visit the URL,
paste the provided code and login using the OIDC flow\&.
The module polls the OpenID Provider token endpoint for information, whether the user has authenticated\&.
Once successful, user is granted access if proper account is requested\&.
For now, only simple mapping described above is available\&.
.br
This mode requires obtaining OAuth 2.0 credentials from OpenID provider of your choice. 
The client ID and client secret must be provided using
.IR client_id
and
.I client_secret
options.

In both of two modes mentioned above, the specified
.I config_url
is used to request OpenID Provider Configuration Information document.
It contains information needed to interact with OP,
including its OAuth 2.0 endpoint locations.

In
.B local_auth
mode, the user is prompted for OIDC access token\&.
The module then sends request to
.I verify_endpoint
motley_cue endpoint
(default being "http://localhost:8080/verify_user")
containing username and access token\&.
User is granted access based on the information in the response\&.

.SH ARGUMENTS
One non-option argument is required\&.
First non-option argument specifies the mode name, the rest is discarded\&.
.br
Following are valid values for the mode name:
.TP
.B id_token
Authenticate using an ID token\&.
The user is prompted for the token\&.
.TP
.B local_auth
Authenticate using an OIDC access token\&.
The user is prompted for the token\&.
.TP
.B poll
Authenticate using the Device Authorization Flow\&.
Requires
.I client_id
and
.IR client_secret \&.


.SH OPTIONS
.TP
.BI --openid-config-url= config_url
OpenID Provider Configuration Information endpoint
(the one with "/\&.well-known/openid-configuration" path)\&.
Default is "https://accounts\&.google\&.com/\&.well-known/openid-configuration".
.br
Applies to
.B id_token
and
.B poll
modes.
.TP
.BI --client_id= client_id
OAuth 2.0 client ID. Required for
.B poll
mode\&.
.TP
.BI --client_secret= client_secret
OAuth 2.0 client secret. Required for
.B poll
mode\&.
.TP
.BI --verify_endpoint= verify_endpoint
motley_cue service endpoint to use for user verification\&.
Default is "http://localhost:8080/verify_user"\&.
.br
Applies to
.B local_auth
mode\&.

.PP
Options that are not valid for selected mode are discarded.

.SH MODULE TYPES PROVIDED
Only
.B auth
is provided\&.


.SH EXAMPLES
.PP
An example usage for /etc/pam\&.d/sshd would be:
.sp
.if n \{\
.RS 4
.\}
.nf
auth    sufficient  pam_oidc\&.so local_auth
.fi
.if n \{\
.RE
.\}


.SH SEE ALSO
.BR pam\&.conf (5),
.BR pam\&.d (5),
.BR pam (8),
.BR sshd (8),
.BR sshd_config (5)


src/pam_oidc.c
#include "../lib/pam_helper.h"
#include "auth.h"
#include <getopt.h>
#include <security/pam_modules.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void promptCallback(const char* prompt, void* context)
{
    char* input;
    getInput((pam_handle_t*)context, 0, &input, prompt);
}

PAM_EXTERN int pam_sm_authenticate(pam_handle_t* pamh, int flags, int argc, const char** argv)
{
    int returnValue;

    const char* username;
    if (returnValue = getUsername(pamh, &username) != PAM_SUCCESS)
        return returnValue;

    fprintf(stderr, "%d args: '", argc);
    for (size_t i = 0; i < argc - 1; i++) {
        fprintf(stderr, "%s ", argv[i]);
    }
    if (argc > 0)
        fprintf(stderr, "%s'\n", argv[argc - 1]);

    char *clientID = NULL, *clientSecret = NULL, *verificationEndpoint = NULL, *openidConfigurationEndpoint = NULL;
    int c;

    while (1) {
        int optionIndex = 0;
        // clang-format off
        static struct option longOptions[] = {
            { "openid-config-url",  required_argument, 0, 4 },
            { "verify-endpoint",    required_argument, 0, 3 },
            { "client-id",          required_argument, 0, 1 },
            { "client-secret",      required_argument, 0, 2 },
            { 0, 0, 0, 0 }
        };
        // clang-format on

        c = getopt_long(argc + 1, (char* const*)(argv - 1), "", longOptions, &optionIndex);
        if (c == -1) {
            fprintf(stderr, "c==-1\n");
            break;
        }

        switch (c) {
        case 1:
            clientID = optarg;
            break;
        case 2:
            clientSecret = optarg;
            break;
        case 3:
            verificationEndpoint = optarg;
            break;
        case 4:
            openidConfigurationEndpoint = optarg;
            break;

        default:
            fprintf(stderr, "?? getopt returned character code 0%o ??\n", c);
            break;
        }
    }

    fprintf(stderr, "optind=%d, argc=%d\n", optind, argc);
    if (--optind >= argc) {
        fprintf(stderr, "No arguments specified - cannot determine mode of operation\n");
        return PAM_AUTH_ERR;
    }

    char* input;

    if (strcmp(argv[optind], "id_token") == 0) {
        getInput(pamh, flags, &input, "ID Token: ");
        return authenticateWithIDToken(username, input, openidConfigurationEndpoint);
    }
    if (strcmp(argv[optind], "local_auth") == 0) {
        getInput(pamh, flags, &input, "Access token: ");
        return authenticateWithMotleyCue(username, input, verificationEndpoint);
    }
    if (strcmp(argv[optind], "poll") == 0) {
        return authenticateWithPolling(
            username, &promptCallback, pamh, openidConfigurationEndpoint, clientID, clientSecret);
    }

    fprintf(stderr, "\nInvalid mode of operation %s\n", argv[optind]);
    return PAM_AUTH_ERR;
}

PAM_EXTERN int pam_sm_setcred(pam_handle_t* pamh, int flags, int argc, const char** argv) { return PAM_SUCCESS; }